### PR TITLE
Added memory sequence minigame

### PR DESCRIPTION
**Pull request by MattWu28**
Added sequence memory mini-game. The layer must select the five randomly chosen tiles in the order shown to succeed. Static green tiles indicate completion while flashing red tiles followed by a repetition of the sequence indicates an incorrect player input.

### CODE REVIEW

File: app/game/memoryGame/index.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                <td><canvas class="canvas" id="canvas3"></canvas></td>
+                <td><canvas class="canvas" id="canvas4"></canvas></td>
+                <td><canvas class="canvas" id="canvas5"></canvas></td>
+            </tr>
+            <tr>
+                <td><canvas class="canvas" id="canvas6"></canvas></td>
+                <td><canvas class="canvas" id="canvas7"></canvas></td>
+                <td><canvas class="canvas" id="canvas8"></canvas></td>
+            </tr>
+        </tbody>
```
Comment (ColinJPage) [inline app/game/memoryGame/index.html:29]: The grid is hard-coded as 3x3, which seems perfectly fine. I'd suggest adding elements these dynamically, but it's not necessary

File: app/game/memoryGame/script.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+const canvas0 = document.getElementById("canvas0"),
+canvas1 = document.getElementById("canvas1"),
+canvas2 = document.getElementById("canvas2"),
+canvas3 = document.getElementById("canvas3"),
+canvas4 = document.getElementById("canvas4"),
+canvas5 = document.getElementById("canvas5"),
+canvas6 = document.getElementById("canvas6"),
+canvas7 = document.getElementById("canvas7"),
+canvas8 = document.getElementById("canvas8");
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:9]: Would there be anything else with id "canvasX"? I think a more descriptive Id could help

File: app/game/memoryGame/script.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+let pattern = [0, 0, 0, 0, 0];
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:11]: Let `patternLength` be a constant

File: app/game/memoryGame/script.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+function renderPattern() {
+    for (let i = 0; i < pattern.length; i++) {
+        setTimeout(() => {
+            const canvas = document.getElementById(`canvas${pattern[i]}`)
+            canvas.style.background = "lightgreen";
+            setTimeout(() => {
+                canvas.style.background = "lightgray";
+            }, 1000);
+        }, 1200 * i);
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:26]: Let's pull these values into named constants at the top of the file.
Something like `blinkDuration` and `timeBetweenBlinks`.
I also think we should lower these values to be pretty quick. The whole minigame should take like 5 seconds ideally

File: app/game/memoryGame/script.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    setTimeout(() => {
+        enableHandler();
+    }, 6000);
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:30]: `timeBetweenBlinks * patternLength`

File: app/game/memoryGame/script.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        canvas.number = i;
+    }
+}
+
+function disableHandler() {
+    for (let i = 0; i < 9; i++) {
+        const canvas = document.getElementById(`canvas${i}`)
+        canvas.removeEventListener("click", handler);
+    }
+}
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:47]: Alternatively, we could leave the event listeners attached, and then in `handler()` choose to ignore the click if the sequence is still playing

File: app/game/memoryGame/script.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+function handler(e) {
+    let curr = e.currentTarget;
+    if (curr.number === pattern[playerPattern.length]) {
+        playerPattern.push(curr.number);
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:52]: Why keep the array `playerPattern`? We could instead keep the index of the current number

File: app/game/memoryGame/script.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        const canvas = document.getElementById(`canvas${i}`);
+        canvas.style.background = "red";
+        setTimeout(() => {
+            canvas.style.background = "lightgray";
+        }, 500);
+    }
+    playerPattern = [];
+    setTimeout(() => {
+        renderPattern();
+    }, 800);
```
Comment (ColinJPage) [inline app/game/memoryGame/script.js:82]: `incorrectPenaltyTime`

File: app/game/memoryGame/index.html [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <link rel="stylesheet" href="styles.css">
+    <script src="script.js" defer></script>
+    <title>Memory Game</title>
+</head>
+<body>
```
Comment (ColinJPage) [inline app/game/memoryGame/index.html:11]: Let's put some message container here. We can swap out the text to let the player know what's going on.
`Memorize the pattern...`
`Repeat the pattern:`
`INCORRECT, try again!`
`Breach resolved!`

**Review by ColinJPage — CHANGES_REQUESTED** [review]
Mostly optional nitpicks. Try experimenting with the constants to increase the pacing of the minigame to like 5-8 seconds

**Review by RichardLam772001 — APPROVED** [review]

**Review by ColinJPage — APPROVED** [review]

**Pull request merged**